### ISSUE TITLE
`s.zero_()` on sparse csr tensor `s` fills values tensor with materialized zeros

### ISSUE THREAD

**Issue by bhosmer** [pytorch/pytorch#82683]
## Issue description

Calling `s.zero_()` on a sparse tensor `s` should clear indices and values. 

Sparse COO tensor does this properly:

```
>>> coo = torch.sparse_coo_tensor(torch.tensor([[0, 1], [2, 3]]), torch.tensor([99, 100]))
>>> coo
tensor(indices=tensor([[0, 1],
                       [2, 3]]),
       values=tensor([ 99, 100]),
       size=(2, 4), nnz=2, layout=torch.sparse_coo)

>>> coo.zero_()
tensor(indices=tensor([], size=(2, 0)),
       values=tensor([], size=(0,)),
       size=(2, 4), nnz=0, layout=torch.sparse_coo)
```

However a sparse CSR tensor will fill its values tensor with zeros instead:
```
>>> csr = torch.sparse_csr_tensor(torch.tensor([0, 2, 4]), torch.tensor([0, 2, 1, 3]), torch.tensor([1, 2, 3, 4]))
>>> csr
tensor(crow_indices=tensor([0, 2, 4]),
       col_indices=tensor([0, 2, 1, 3]),
       values=tensor([1, 2, 3, 4]), size=(2, 4), nnz=4,
       layout=torch.sparse_csr)
>>> csr.zero_()
tensor(crow_indices=tensor([0, 2, 4]),
       col_indices=tensor([0, 2, 1, 3]),
       values=tensor([0, 0, 0, 0]), size=(2, 4), nnz=4,
       layout=torch.sparse_csr)
```

I think the implementation comes from [here](https://github.com/pytorch/pytorch/pull/73075), which AFAICT uses a macro for delegating pointwise ops to the values tensor in the CSR case. 

Should be relatively easy to implement it properly following the coo example [trailhead](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorFactories.cpp#L228)

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer - I'm wildly guessing this should probably land in one of your queues, but LMK if not. in any event seems worth fixing for 1.13 _[edit: lol the tag-to-cc machine found my ats of Andrew and Nik and added everybody else to them]_

cc @zdevito who reported

**Comment by pearu** [comment]
> Calling `s.zero_()` on a sparse tensor should clear indices and values.

@bhosmer There appears to be autograd related reason why @IvanYashchuk implemented `csr.zero_()` semantics as `csr.values().zero_()`. I'm investigating this atm, so, stay tuned...

**Comment by IvanYashchuk** [comment]
Why is it useful to drop the indices and why do you think the COO tensor does this properly and CSR's behavior is wrong?

**Comment by pearu** [comment]
> Why is it useful to drop the indices and why do you think the COO tensor does this properly and CSR's behavior is wrong?

The same question can be asked the other way around: why not drop indices and why CSR behavior is correct?

The answers depend on the use cases.

Here follows some reasoning points:
- From the sparse semantics point of view (read: sparse formats are implementation details), a tensor `x` after `x.zero_()` should represent a tensor whose elements are all zeros. Both possible implementations (`x.values().zero_()` and "reset x.nnz to zero") fulfill this property.
- From the performance point of view, "reset x.nnz to zero" has an advantage of storage efficiency and also of indirect processor efficiency (no need to process values that are zero anyway).

IIUC, the current CSR's behavior was implemented because of
https://github.com/pytorch/pytorch/blob/ebd4e90ff7f6d1e57bbf6f2e717a8addb6b28e28/torch/csrc/autograd/FunctionsManual.cpp#L1248-L1249
where `zeros_like` is expected to return a zero tensor with the same indices as `mat1`, otherwise, autograd on CSR tensors would be broken (as `addmm` would always return zero tensor if its first input has nnz==0). In #85030 , this code is replaced with an equivalent code
https://github.com/pytorch/pytorch/blob/fd22b3e57daa5c444cad12d18cf2edd41fe81a93/torch/csrc/autograd/FunctionsManual.cpp#L1248-L1251
and with that, we have the freedom to choose a more memory-efficient implementation of `zeros_like`, `zeros`, etc that underhood all use `zero_`.
